app/src/main/java/com/xmartlabs/template/ui/RecyclerViewGridSpacingDecoration.java

```diff
@@ -0,0 +1,253 @@
+   * @param view the view to add the spacing
+   * @param recyclerView the recycler view that holds the {@code view}
+   * @param gridLayoutManager the layout manager of the recycler view
+   */
+  private void setOffsetForItem(Rect outRect, View view, RecyclerView recyclerView, GridLayoutManager gridLayoutManager) {
+    int position = recyclerView.getChildLayoutPosition(view);
+    int spanCount = gridLayoutManager.getSpanCount();
+    int numberOfItems = recyclerView.getAdapter().getItemCount();
+    GridLayoutManager.SpanSizeLookup spanSizeLookup = gridLayoutManager.getSpanSizeLookup();
+
+    Optional.ofNullable(setItemOffsetConsumer)
+        .executeIfPresent(consumer -> consumer.accept(outRect, recyclerView))
+        .executeIfAbsent(() -> {
+          int firstRowTopSpacing = Optional.ofNullable(this.firstRowTopSpacing).orElse(itemSpacing);
+          int firstColumnLeftSpacing = Optional.ofNullable(this.firstColumnLeftSpacing).orElse(itemSpacing);
+          int lastColumnRightSpacing = Optional.ofNullable(this.lastColumnRightSpacing).orElse(itemSpacing);
+          int lastRowBottomSpacing = Optional.ofNullable(this.lastRowBottomSpacing).orElse(itemSpacing);
+
+          outRect.top = isFirstRow(position, spanCount, spanSizeLookup) ? firstRowTopSpacing : itemSpacing;
+          outRect.left = isFirstColumn(position, spanCount, spanSizeLookup) ? firstColumnLeftSpacing : itemSpacing;
+          outRect.right = isLastColumn(position, spanCount, numberOfItems, spanSizeLookup) ? lastColumnRightSpacing : itemSpacing;
+          outRect.bottom = isLastRow(position, spanCount, numberOfItems, spanSizeLookup) ? lastRowBottomSpacing : itemSpacing;
+        });
+  }
+
+  /**
+   * Calculates whether or not the item at {@code position} belongs to the first row.
+   *
+   * @param position the item position
+   * @param spanCount the maximum number of items a row can hold
+   * @param spanSizeLookup the object that defines how much space an item can take
+   * @return whether or not the item belong to the first row
+   */
+  private boolean isFirstRow(int position, int spanCount, @NonNull GridLayoutManager.SpanSizeLookup spanSizeLookup) {
+    return spanSizeLookup.getSpanGroupIndex(position, spanCount) == FIRST_ROW_GROUP;
+  }
+
+  /**
+   * Calculates whether or not the item at {@code position} belongs to the last row.
+   *
+   * @param position the item position
+   * @param spanCount the maximum number of items a row can hold
+   * @param numberOfItems the total number of items held by the {@link RecyclerView}
+   * @param spanSizeLookup the object that defines how much space an item can take
+   * @return whether or not the item belongs to the last row
+   */
+  private boolean isLastRow(int position, int spanCount, int numberOfItems,
+                            @NonNull GridLayoutManager.SpanSizeLookup spanSizeLookup) {
+    while (position < numberOfItems - 1) {
```

Comment:
I don't think this particular while qualifies to be used with Stream, since it's not iterating through a collection and the number of iterations is not known